### ISSUE TITLE
Rewrite twisted.web.static.File.directoryListing to not use woven

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3257]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3257|
|Type|enhancement|
|Created|2008-05-30 15:17:40Z|
|Branch|https://github.com/twisted/twisted/tree/web-dirlist-3257-2|

woven should be deleted at some point. The current one is not particularly awesome too. We should have a look at nevow's one, even if it's bad too. bad + bad = great!

<details><summary>Searchable metadata</summary>

```
trac-id__3257 3257
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_web_dirlist_3257_2 branches/web-dirlist-3257-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1212160660000000 1212160660000000
changetime__1214939450000000 1214939450000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23795]) Branching to 'web-dirlist-3257'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23796]) Reimplement DirectoryLister without woven, cleanups.

Refs [#3257](https://github.com/twisted/twisted/issues/3257)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [24134]) Branching to 'web-dirlist-3257-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

Yay, woven is almost gone.

`formatFileSize` has a typo, _uman_.  Also, is the bytes suffix intentionally the only lower-case suffix?

`DirectoryLister` is missing some tests, I think.  eg, the `dirs` parameter is never passed in the tests.  Testing would probably be simplified a bunch if `render` got split into a few methods (eg, one for producing a list of filename/file info tuples, another for iterating over that and producing markup, etc).

The `pattern` attribute in the markup in `linePattern` isn't used for anything.

`FilePath` might make the code a little simpler?  Or maybe the need for compatibility would just result in it being more complex with `FilePath`, I'm not sure.

`os.stat` returns an object with an `st_size` attribute that gives the same result as getting the `stat.ST_SIZE`'th element from it.

Some quoting isn't quite correct - eg, for the text of the link for each path, html entities aren't quoted, so '&' ends up in the page as a literal '&' instead of as '&amp;'.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

(In [24157])  * Escape &/</>
 * use st_size attribute
 * remove useless pattern attribute
 * uppercase bytes

Refs [#3257](https://github.com/twisted/twisted/issues/3257)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve removed owner|
|-|-|

Thanks for the review! Everything addressed, I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

  * broken symlinks cause an unhandled exception
  * `DirectoryLister.render` should have a docstring, and the attributes should be documented in the class doc string
  * missing whitespace around exponentiation in `formatFileSize`

Also, I notice this version is about 15x faster than the woven version.  Just thought you might want to know. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Replying to [exarkun](#note_8):

> broken symlinks cause an unhandled exception

Done.

> `DirectoryLister.render` should have a docstring, and the attributes should be documented in the class doc string

Done.

>  missing whitespace around exponentiation in `formatFileSize`

Done.

> Also, I notice this version is about 15x faster than the woven version.  Just thought you might want to know. :)

It's even better when we tackle some bugs at the same time :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Some buildslaves had a failure:

```
[FAIL]: twisted.web.test.test_static.DirectoryListerTest.test_mimeTypeAndEncodings

Traceback (most recent call last):
  File "/Users/buildbot/Run/buildbot/full2.4-exarkun-osx/Twisted/twisted/web/test/test_static.py", line 258, in test_mimeTypeAndEncodings
    self.assertEquals(files, [
twisted.trial.unittest.FailTest: not equal:
a = [{'encoding': '',
  'href': 'file1.txt',
  'size': '5B',
  'text': 'file1.txt',
  'type': '[text/plain]'},
 {'encoding': '',
  'href': 'file2.py',
  'size': '6B',
  'text': 'file2.py',
  'type': '[text/plain]'},
 {'encoding': '[gzip]',
  'href': 'file3.conf.gz',
  'size': '15B',
  'text': 'file3.conf.gz',
  'type': '[text/plain]'},
 {'encoding': '[bzip2]',
  'href': 'file4.diff.bz2',
  'size': '15B',
  'text': 'file4.diff.bz2',
  'type': '[text/plain]'}]
b = [{'encoding': '',
  'href': 'file1.txt',
  'size': '5B',
  'text': 'file1.txt',
  'type': '[text/plain]'},
 {'encoding': '',
  'href': 'file2.py',
  'size': '6B',
  'text': 'file2.py',
  'type': '[text/x-python]'},
 {'encoding': '[gzip]',
  'href': 'file3.conf.gz',
  'size': '15B',
  'text': 'file3.conf.gz',
  'type': '[text/plain]'},
 {'encoding': '[bzip2]',
  'href': 'file4.diff.bz2',
  'size': '15B',
  'text': 'file4.diff.bz2',
  'type': '[text/x-diff]'}]
```

Seems to be related to the `.diff.bz2` extension.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Also, on Windows:

```
[ERROR]: twisted.web.test.test_static.DirectoryListerTest.test_brokenSymlink

Traceback (most recent call last):
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\web\test\test_static.py", line 290, in test_brokenSymlink
    file1.linkTo(path.child("file2"))
  File "s:\buildbots\twisted\server-2k8-x86-py2.5-select\Twisted\twisted\python\filepath.py", line 426, in linkTo
    os.symlink(self.path, linkFilePath.path)
exceptions.AttributeError: 'module' object has no attribute 'symlink'
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [24166]) Fix buildbot failures

Refs [#3257](https://github.com/twisted/twisted/issues/3257)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

Ok buildbot seems happier.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

I guess there should be a couple direct unit tests for `formatFileSize`.  Other than that, looks good.  Merge when you're satisfied.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `closed`|
|-|-|

(In [24203]) Merge web-dirlist-3257-2

Author: therve
Reviewer: exarkun
Fixes [#3257](https://github.com/twisted/twisted/issues/3257)

Rewrite twisted.web directory lister without woven.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
